# NHTSA API SQL Analysis

This notebook contains descriptive and diagnostic analytics queries against the `nhtsa_recalls_summary` and `nhtsa_recalls_detail` tables.

## Setup
Load credentials from environment variables and establish a database connection.

In [1]:
import os
import pandas as pd
from sqlalchemy import create_engine
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

PG_user = os.getenv('PG_USER')
PG_password = os.getenv('PG_PASSWORD')
PG_host = os.getenv('PG_HOST')
PG_DB = os.getenv('PG_DB')

# Create database engine
engine = create_engine(f"postgresql://{PG_user}:{PG_password}@{PG_host}/{PG_DB}")

# Display all rows in pandas DataFrame
pd.set_option('display.max_rows', None)


## Query 1: Descriptive Analytics Query
**Business Question:** Which makes and models have the highest number of recalls?

In [2]:
sql_query = '''
-- Descriptive Analytics Query for NHTSA Recalls (with proper quoting)
SELECT
    "Make"            AS make,
    "Model"           AS model,
    "RecallCount_2000_2025" AS recall_count
FROM sql_project."nhtsa_recalls_summary"
ORDER BY recall_count DESC;
'''
df1 = pd.read_sql(sql_query, engine)
df1

make                    model  recall_count
0   Toyota                   Tundra           218
1   Toyota                   Tacoma           151
2   Toyota                   Sienna           148
3   Toyota                  Corolla           145
4   Toyota                     RAV4           130
5   Toyota                  Sequoia           128
6   Toyota                  4Runner           120
7   Toyota               Highlander           113
8   Toyota                    Camry           103
9   Toyota                    Prius            84
10  Toyota                   Avalon            81
11  Toyota                    Yaris            68
12  Toyota               FJ Cruiser            66
13  Toyota             Land Cruiser            52
14  Toyota        Highlander Hybrid            45
15  Toyota                    Venza            35
16  Toyota             Camry Hybrid            29
17  Toyota                   Matrix            27
18  Toyota            Avalon Hybrid            18
19  Toyota            Tundra Hybrid            16
20  Toyota                  Prius v            15
21  Toyota                    Supra            12
22  Toyota              RAV4 Hybrid            11
23  Toyota                  RAV4 EV             9
24  Toyota           Corolla Hybrid             7
25  Toyota               RAV4 Prime             7
26  Toyota                   Celica             7
27  Toyota              Prius Prime             7
28  Toyota                     Echo             7
29  Toyota            Sienna Hybrid             6
30  Toyota            Corolla Cross             6
31  Toyota     Corolla Cross Hybrid             6
32  Toyota                     C-HR             6
33  Toyota                  Prius c             5
34  Toyota                    Crown             4
35  Toyota                       86             4
36  Toyota        Corolla Hatchback             3
37  Toyota                     GR86             3
38  Toyota                     bZ4X             3
39  Toyota                    Mirai             3
40  Toyota         Grand Highlander             2
41  Toyota  Grand Highlander Hybrid             2
42  Toyota               GR Corolla             2
43  Toyota                 GR Supra             1
44  Toyota              Yaris Verso             0
45  Toyota                   Aurion             0
46  Toyota                    Auris             0
47  Toyota                   Avanza             0
48  Toyota                  Avensis             0
49  Toyota                    Carib             0
50  Toyota                   Carina             0
51  Toyota                 Cavalier             0
52  Toyota            Corolla Verso             0
53  Toyota                 Cressida             0
54  Toyota                     Dyna             0
55  Toyota                    Etios             0
56  Toyota               Etios Liva             0
57  Toyota                     FJ40             0
58  Toyota                 Fortuner             0
59  Toyota                  Harrier             0
60  Toyota                    Hiace             0
61  Toyota                    Hilux             0
62  Toyota                   Innova             0
63  Toyota                      iQ2             0
64  Toyota                 King Cab             0
65  Toyota                   Mark X             0
66  Toyota                      MR2             0
67  Toyota                    Paseo             0
68  Toyota                   Pickup             0
69  Toyota                    Prado             0
70  Toyota                   Premio             0
71  Toyota                   Previa             0
72  Toyota            Prius Plug-in             0
73  Toyota                     Raum             0
74  Toyota                    Scion             0
75  Toyota                       Se             0
76  Toyota                   Solara             0
77  Toyota                 Sprinter             0
78  Toyota                     T100             0
79  To

### Insight

Five Toyota nameplates (Tundra 218, Tacoma 151, Sienna 148, Corolla 145, RAV4 130) make up ~41 % of all Toyota recalls

### Recommendation

Place these models at the top of the recall-reduction backlog: build a quick dashboard that tracks their open campaigns and highlights any spike > 5 recalls/quarter.

### Prediction

Concentrated effort on just these five programs could trim Toyota’s annual recall volume by a large amount.

## Query 2: Diagnostic Analytics Query
**Business Question:** For the top 5 recalled models, which components are most frequently recalled and what percentage do they represent?

In [ ]:
sql_query = '''
WITH top_models AS (
  SELECT
    "Make",
    "Model",
    "RecallCount_2000_2025" AS recall_count
  FROM sql_project."nhtsa_recalls_summary"
  ORDER BY recall_count DESC
  LIMIT 5
),
component_counts AS (
  SELECT
    "Manufacturer",
    "Model",
    "Component",
    COUNT(*) AS component_count
  FROM sql_project."nhtsa_recalls_detail"
  GROUP BY "Manufacturer", "Model", "Component"
)
SELECT
  tm."Make"  AS make,
  tm."Model" AS model,
  cc."Component"          AS component,
  cc.component_count      AS count,
  cc.component_count * 100.0
        / SUM(cc.component_count) OVER (PARTITION BY tm."Make", tm."Model")
        AS pct_of_model_recalls
FROM top_models tm
JOIN component_counts cc
  ON LOWER(cc."Model") = LOWER(tm."Model")
 AND LOWER(cc."Manufacturer") ILIKE '%%' || LOWER(tm."Make") || '%%'   -- note the doubled %%
ORDER BY tm."Model", count DESC;
'''
df2 = pd.read_sql(sql_query, engine)
df2



make    model                                          component  \
0    Toyota  Corolla    AIR BAGS:FRONTAL:PASSENGER SIDE:INFLATOR MODULE   
1    Toyota  Corolla                                           AIR BAGS   
2    Toyota  Corolla         AIR BAGS: AIR BAG/RESTRAINT CONTROL MODULE   
3    Toyota  Corolla                             EQUIPMENT:OTHER:LABELS   
4    Toyota  Corolla                             EQUIPMENT:OTHER:LABELS   
5    Toyota  Corolla            VEHICLE SPEED CONTROL:ACCELERATOR PEDAL   
6    Toyota  Corolla                          ENGINE AND ENGINE COOLING   
7    Toyota  Corolla            SEATS:FRONT ASSEMBLY:SEAT HEATER/COOLER   
8    Toyota  Corolla       VISIBILITY:POWER WINDOW DEVICES AND CONTROLS   
9    Toyota  Corolla    AIR BAGS:FRONTAL:SENSOR/CONTROL MODULE-INACTIVE   
10   Toyota  Corolla           FUEL SYSTEM, GASOLINE:DELIVERY:FUEL PUMP   
11   Toyota  Corolla                             EQUIPMENT:OTHER:LABELS   
12   Toyota  Corolla                                           STEERING   
13   Toyota  Corolla       AIR BAGS:FRONTAL:DRIVER SIDE:INFLATOR MODULE   
14   Toyota  Corolla            AIR BAGS:SENSOR:OCCUPANT CLASSIFICATION   
15   Toyota  Corolla                                   AIR BAGS:FRONTAL   
16   Toyota  Corolla                    STRUCTURE:BODY:ROOF AND PILLARS   
17   Toyota  Corolla                                    STEERING:COLUMN   
18   Toyota  Corolla                             EQUIPMENT:OTHER:LABELS   
19   Toyota  Corolla      SERVICE BRAKES, HYDRAULIC:POWER ASSIST:VACUUM   
20   Toyota  Corolla                        VISIBILITY:GLASS, SIDE/REAR   
21   Toyota  Corolla             SEAT BELTS: REAR/OTHER:BUCKLE ASSEMBLY   
22   Toyota  Corolla                          POWER TRAIN:AXLE ASSEMBLY   
23   Toyota  Corolla               TIRES:TEMPORARY/EMERGENCY SPARE TIRE   
24   Toyota  Corolla                                             WHEELS   
25   Toyota  Corolla   VISIBILITY:WINDSHIELD WIPER/WASHER:SWITCH/WIRING   
26   Toyota  Corolla  POWER TRAIN:AUTOMATIC TRANSMISSION:TORQUE CONV...   
27   Toyota  Corolla                             EQUIPMENT:OTHER:LABELS   
28   Toyota     RAV4                                    SUSPENSION:REAR   
29   Toyota     RAV4            SEATS:FRONT ASSEMBLY:SEAT HEATER/COOLER   
30   Toyota     RAV4       VISIBILITY:POWER WINDOW DEVICES AND CONTROLS   
31   Toyota     RAV4            VEHICLE SPEED CONTROL:ACCELERATOR PEDAL   
32   Toyota     RAV4                              SEAT BELTS:REAR/OTHER   
33   Toyota     RAV4              ELECTRICAL SYSTEM:12V/24V/48V BATTERY   
34   Toyota     RAV4                             EQUIPMENT:OTHER:LABELS   
35   Toyota     RAV4                             EQUIPMENT:OTHER:LABELS   
36   Toyota     RAV4       AIR BAGS:FRONTAL:DRIVER SIDE:INFLATOR MODULE   
37   Toyota     RAV4                             EQUIPMENT:OTHER:LABELS   
38   Toyota     RAV4              STEERING:ELECTRIC POWER ASSIST SYSTEM   
39   Toyota     RAV4                                   VISIBILITY/WIPER   
40   Toyota     RAV4    AIR BAGS:FRONTAL:SENSOR/CONTROL MODULE-INACTIVE   
41   Toyota     RAV4                             EQUIPMENT:OTHER:LABELS   
42   Toyota     RAV4                                           AIR BAGS   
43   Toyota     RAV4                                    TRAILER HITCHES   
44   Toyota     RAV4             SUSPENSION:FRONT:CONTROL ARM:LOWER ARM   
45   Toyota     RAV4    AIR BAGS:FRONTAL:PASSENGER SIDE:INFLATOR MODULE   
46   Toyota     RAV4           FUEL SYSTEM, GASOLINE:DELIVERY:FUEL PUMP   
47   Toyota     RAV4                                          EQUIPMENT   
48   Toyota     RAV4            AIR BAGS:SENSOR:OCCUPANT CLASSIFICATION   
49   Toyota     RAV4                          ENGINE AND ENGINE COOLING   
50   Toyota     RAV4                                        TIRES:VALVE   
51   Toyota     RAV4                     SERVICE BRAKES, AIR:DISC:ROTOR   
52   Toyota     RAV4            

In [ ]:
df2.groupby(['model']).apply(
    lambda g: g.nlargest(1, 'pct_of_model_recalls')[['component',
                                                     'pct_of_model_recalls']]
)


/tmp/ipykernel_18341/4126860498.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df2.groupby(['model']).apply(


component  \
model                                                          
Corolla 0    AIR BAGS:FRONTAL:PASSENGER SIDE:INFLATOR MODULE   
RAV4    28                                   SUSPENSION:REAR   
Sienna  64           SEATS:FRONT ASSEMBLY:SEAT HEATER/COOLER   
Tacoma  99           SEATS:FRONT ASSEMBLY:SEAT HEATER/COOLER   
Tundra  139           STEERING:HYDRAULIC POWER ASSIST SYSTEM   

             pct_of_model_recalls  
model                              
Corolla 0               29.310345  
RAV4    28              13.846154  
Sienna  64              10.769231  
Tacoma  99               8.633094  
Tundra  139             15.306122

### Insight

Each high-recall model shows a single dominant component family—e.g., passenger air-bag inflators drive 29 % of Corolla recalls, rear-suspension links 14 % of RAV4, hydraulic power-assist 15 % of Tundra.

### Recommendation

Run root-cause analyses starting with the top component for each model; prioritize supplier audits, design reviews, or process-control checks on those items first.

### Prediction

Eliminating the leading component issue per model would cut total recalls for that vehicle line, generating potential savings.